### DISCUSSION TITLE
Welcoming? Successor language?

### DISCUSSION THREAD

**Discussion by jmafc** [carbon-language/carbon-lang#1961] [Language design]
Please refer to issue #1948. I will copy my original and subsequent posts in case @cjdb decides to take further unwelcoming steps.  First of all, I'd like to offer my opinion on what was unwelcome about the actions. If you run an open source project and *emphasize* that you want to create a welcoming community, and someone posts something in your issue tracker which you consider is not the "right" place to start a discussion, I think *any* community member, even the founder or leader, in order to *be* welcoming would make a comment letting the poster know that he/she should move the discussion to the "appropriate" other channel(s). Then, if the poster didn't act on the suggestion in a "reasonable" time (considering the global timezones involved, I think at least 24 hours would be reasonable), then perhaps closing the issue may be appropriate. I consider that closing the issue *immediately* as "not planned" is *not* welcoming. Locking the issue to further input by the poster and two *other*  participants simply adds to the unwelcoming attitude. Saying "thanks for your input" is *not* sufficient/hypocritical when the obvious message is "go away!"

---

Some mostly random thoughts (after watching the announcement video, including Q&A, and browsing some issues):

Perhaps the community is welcoming, but have you considered that the syntax, albeit easier to parse, may not be welcoming to the thousands of C/C++ developers who are used to type name and would see let (or var) name: type as a retrogression to Pascal or BASIC? Even if you provide a tool to convert existing code, the new code will have to be written using the new syntax, so arguably that's an unwelcome change.

I'm afraid that claiming Carbon is a kind of "successor" language is stretching the analogies a bit. C++ of course kept 99% of the C syntax. If I'm not mistaken, Typescript did not introduce a radical new syntax for declaring entities, only type annotations. I'm not familiar with Kotlin, but it doesn't appear that different from Java.

The example in the README uses C++ #include's rather than the more modern module syntax. That is somewhat similar to the way initial Rust presentations would emphasize how "atrociously" C++ behaved with examples using raw pointers instead of unique or shared pointers (I realize most existing C++ code uses #include's, but it seems fairer to do a comparison using the newer syntax).

Last but certainly not least is the issue of error handling. I could not find a clear position but it seems that Carbon may not have try/catch. If so, is that seen as an improvement? How can Carbon then be a "successor" to C++?

---
Forgot to mention that there is already a "successor" language to C++ that, IIRC, has kept most of the C/C++ syntax intact (except for arrays). It's Walter Bright's D. It has modules and even design by contract features.

---
@cabmeurer: I can understand your thoughts on this, and would like to share some of mine. The details of chosen syntax you described as retrogression, I feel like is actually much more common for modern languages/design. With that in mind, I think the "thousands of C/C++ developers" that are use to type name are bound to be familiar enough with other languages like Kotlin, Go, Swift and many other languages that use the let (or var) name: type syntax.
I personally feel the mission for Carbon is to be a successor to C++, however whether or not you agree with that is a different thing. I have had several conversations with different devs, and some of their opinions are that rather than create a new language, that the efforts should be made to make another language the 'successor' (I've heard people say this about Rust, Go and even Swift) as of now, the language is experimental and time will tell.
As far as error handling, if we take a look at the docs, [Roadmap](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/roadmap.md) it is stated that version 0.1 isn't slated to be released until the end of this year, and 1.0 isn't planned until 2024/2025. To quote the Roadmap:
"The design choices made to reach this point are expected to be experimental, and many of them may need revisiting before we reach 1.0, but the broad shape of the language should be clear at this point, and it should be possible to write non-trivial Carbon programs." So I would imagine something like try/catch would be reasonable to expect in the future.

---
It's one thing to know about let/var name: type-style syntax, it's quite another to be agreeable to code in it day in and day out. Rust, Go and Pascal's declaration syntax may be "modern", but I wouldn't want to use it on a daily basis.
"Successor" in what sense? C++ is a successor to C because you can take a large software base written in C, rename the .cs to .cpps (or whatever is your preferred C++ extension) and 99% of the files will compile (and you may get better warnings). I haven't actively looked at D recently but with what they called -BetterC you could compile a C file directly. Carbon at this moment is vaporware. At least Go, when it was announced outside of Google, had an implementation (perhaps written in C++, but nevertheless something "tangible") and although they didn't have generics and their build system went through multiple iterations after their first release, the overall syntax and language semantics were about 80% settled.
Plus one needs to consider that C++ will not remain static until 2024/2025 (that's why any comparisons ought to be against C++20 with modules and even what's been announced for C++23).
I'm sorry but your quote from the Roadmap doesn't inspire confidence.

---
@davidhunter22: In my personal opinion Carbon should drop any talk of "Successor to C++". As a long time C++ prgrammer I would love a better C++ but it seems very arrogant to claim to be that from the start, to the point where I am put of even trying the language. I would much prefer a more humble approach, for instance just focus on why if your a developer/company doing C++ why Carbon may be better than continuing to use C++ or switching to D or Rust and why migrating to it is easy.

Also I would stress the Visual Studio integration. The entire effort will always be a niche, like D, without this.

---
@cjdb: Thanks for your input! This is more of a discussion than an issue, so I'm going to close this. Please see [Where should I ask questions about Carbon Language?](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/faq.md#where-should-i-ask-questions-about-carbon-language) from our FAQ.

I will address some points very briefly, but please use the appropriate channels instead of responding here.

    Perhaps the community is welcoming, but have you considered that the syntax, albeit easier to parse, may not be welcoming to the thousands of C/C++ developers who are used to type name and would see let (or var) name: type as a retrogression to Pascal or BASIC? Even if you provide a tool to convert existing code, the new code will have to be written using the new syntax, so arguably that's an unwelcome change.

The Carbon Language declaration syntax is to counter C++'s [most vexing parse issue](https://en.wikipedia.org/wiki/Most_vexing_parse).

    I'm afraid that claiming Carbon is a kind of "successor" language is stretching the analogies a bit. C++ of course kept 99% of the C syntax. If I'm not mistaken, Typescript did not introduce a radical new syntax for declaring entities, only type annotations. I'm not familiar with Kotlin, but it doesn't appear that different from Java.

Kotlin has diverged from Java's syntax in some fairly significant ways, and a lot of those share similarity with the changes proposed in Carbon. You can see some of those in the [Kotlin documentation](https://kotlinlang.org/docs/basic-syntax.html).

    The example in the README uses C++ #include's rather than the more modern module syntax. That is somewhat similar to the way initial Rust presentations would emphasize how "atrociously" C++ behaved with examples using raw pointers instead of unique or shared pointers (I realize most existing C++ code uses #include's, but it seems fairer to do a comparison using the newer syntax).

[Carbon is currently targeting C++17 as its entry-point for interop](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/interoperability/philosophy_and_goals.md#interoperability-with-and-migration-from-existing-c-code), so C++20 modules are unavailable. That isn't to say that C++20 and beyond won't receive interop support, but it's not a priority in the initial roadmap.

    Last but certainly not least is the issue of error handling. I could not find a clear position but it seems that Carbon may not have try/catch. If so, is that seen as an improvement? How can Carbon then be a "successor" to C++?

[Carbon is still very much in design](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/faq.md#how-complete-is-carbons-design). A lack of exception handling in the specification doesn't necessarily mean that there won't be exceptions: it just means that no one has looked into them just yet.

    Forgot to mention that there is already a "successor" language to C++ that, IIRC, has kept most of the C/C++ syntax intact (except for arrays). It's Walter Bright's D. It has modules and even design by contract features.

Please see [What alternatives did you consider? Why did they not work?](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/faq.md#what-alternatives-did-you-consider-why-did-they-not-work). https://github.com/carbon-language/carbon-lang/discussions/1448 is our ticket to motivate someone writing an explicit section for why we don't think D works; you might like to subscribe to it so you get a notification once the FAQ is updated.

---
0.  Well, you have proven what I was going to write before: "It is very easy to say that you're going to have a "welcoming community", but the proof is in the pudding". Closing the issue as "not planned" is not welcoming. BTW, if you click the question you mention in the FAQ (not the one in your text above) it doesn't redirect properly. Since I'm not welcome, I will not bother with GitHub discussions.
1.  Have you read the "Most vexing parse issue" article? It says "the phenomenon was quite common in C++ until the introduction of uniform initialization in C++11". What is the point of referring to that now? Also, the first example is about a C-style cast? How many conscientious C++ programmers use C-style casts in their code in 2022?
2.  I don't know much (or care) about Kotlin, but I don't think JetBrains developed it claiming it to be a "successor" to Java. Isn't Scala also a "successor" in that sense. I also agree with David Hunter suggestion that Carbon should drop its C++ succession claim (it sounds like princes battling it out for a throne).
3.  The point is not about interoperability with some version of C++. My point is that if you're going to try to convince someone that something is better (or is going to be better) than something else, you should be making realistic comparisons.
4.  Ah, but someone has looked into exceptions. The reason I brought it up, is because of https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/principles/error_handling.md, which states "Carbon errors, unlike exceptions in C++ and similar languages, will not be propagated implicitly" (emphasis added). How can Carbon expect to interoperate with C++ if, in essence, it wants the programmer to pass the errors "up the chain" explicitly?

**Comment by chandlerc** [comment]
It seems like there are two high-level issues raised here -- there is a huge amount of text, so I'm not going to be able to respond point by point, but I'll try to address these two high level issues.

However, I want to remind you to avoid using sarcasm to make your points on this forum. I understand that you feel strongly about these issues, but being sarcastic or snarky is not productive or effective and not something we do here.

----

First, is a concern around whether the community is "welcoming" or not.

=/ It's a hard balance to make a welcoming community, and impossible to be welcoming for everyone.

Not every moderation action is perfect or without flaw. For example, we're still learning how to use GitHub most effectively -- hopefully in the future we can transfer issues to discussions rather than closing them. But we are trying to make sure our issues stay focused on specific changes and the discussion forums are used for more open-ended discussion. We're doing the best we can to keep that, as that is super important for others to feel welcome to participate and be productive. While the action was taken very quickly here, nothing was deleted or lost forever. You were able to move the discussion here which seems like a more appropriate forum. So while a little inefficient, this seems fine to me.

Similarly, we try to have an FAQ to be helpful, but sometimes links don't work as well as we'd like or we have a typo.

I don't think calling this "unwelcoming" is really helpful or moving the discussion forward. It both holds people to an unreasonable standard of not making simple mistakes, and ignores that there are tradeoffs here. We can't be perfectly welcoming to all people at all times.

----

The second is that there are a number of design decisions and strategy decisions that you don't agree with.

On the design decisions, @cjdb pointed at several pieces of information that informed these decisions. You may still not agree, these are complex decisions and it is very likely that reasonable people will disagree on them. But the important question is whether there is *specific new information* that needs to be considered. Otherwise, we've made the best technical judgement call we can given the information, and we should move forward with that.

On the strategic side, we've chosen to articulate Carbon's strategy as a "successor" strategy -- but this is *new* terminology and we've tried to define it. Maybe that wording doesn't work for everyone, but picking wording and terminology like this is extremely hard. We've done the best we can, and just saying "don't use this term" doesn't help us. If there is some specific problem with this approach, combined with a constructive suggestion for another approach, I would focus on that.

**Reply by OlaFosheimGrostad** [reply]
> On the strategic side, we've chosen to articulate Carbon's strategy as a "successor" strategy -- but this is _new_ terminology and we've tried to define it.

The wording does not work because "successor" already has an established interpretation so you cannot use it to create new terminology. It usually means a second/third/fourth/… attempt by the original community at capturing some key language qualities in their next language effort, correcting shortcoming of their previous attempt. E.g. BETA was the successor for Simula, but is completely different, yet has similar modelling goals. There is no compatibility involved in being a successor. Java is semantically very very close to Simula, but it wasn't really a successor by intent or culture, despite being almost a replica. Neither was C++.

**Reply by chandlerc** [reply]
> > On the strategic side, we've chosen to articulate Carbon's strategy as a "successor" strategy -- but this is _new_ terminology and we've tried to define it.
> 
> The wording does not work because "successor" already has an established interpretation so you cannot use it to create new terminology. It usually means a second/third/fourth/… attempt by the original community at capturing some key language qualities in their next language effort, correcting shortcoming of their previous attempt. E.g. BETA was the successor for Simula, but is completely different, yet has similar modelling goals. There is no compatibility involved in being a successor. Java is semantically very very close to Simula, but it wasn't really a successor by intent or culture, despite being almost a replica. Neither was C++.

While this is your interpretation of "successor", others (including both myself and numerous reviewers who helped us craft the current phrasing) have different interpretations. I don't think there is a *single* established interpretation.

The point of us clarifying what we mean with the term is to address the fact that there are different interpretations. I'm not aware of a useful term that wouldn't have this problem, and so wouldn't need this kind of clarification.

**Reply by OlaFosheimGrostad** [reply]
> While this is your interpretation of "successor", others (including both myself and numerous reviewers who helped us craft the current phrasing) have different interpretations. I don't think there is a _single_ established interpretation.

Well, this isn't really a big issue, but it was a rather clever marketing move to call it a "C++ successor", yet I am not the only one that also thought that it takes a lot of *hubris* to do so. :-) So, overall a good move on your part, you got the attention to kick-start your project (being slightly controversial makes for better headlines).

For people growing up in a monarchy the power-element associated with "succession" is quite clear though. This is also a big factor in how languages are perceived. Kotlin was just another bland JVM language until the Android-aristocracy dethroned Java. Nobody would have given Swift another look if the Apple-aristocracy hadn't installed it as the new king. And so on… On the other hand Dart was intended to be the installed successor for JavaScript, but the Chrome-aristocracy chose not to, leaving Dart in limbo for a long time.

TypeScript is not succeeding JavaScript though, it is more of a piggy-back/parasitic/symbiotic approach. TypeScript is a typed JavaScript *dialect*,  Objective-C and C++ are  C dialects…

Carbon is different from these, in my view. It seems more like you try to go for a grass-root revolution by the people (users) shredding the conservative structures that govern C++. Which could be excellent if it was to happen, but also quite different dynamics from the ones you listed as "successor languages"? Carbon isn't being parasitic, but takes more of a conversion/transpiling approach.

TLDR; forget about the "successor-language" rhetoric, it does not seem to be a good fit, embrace instead the fact that you are not doing another round of evolution of C++, but take lessons from it.

**Comment by jmafc** [comment]
@chandlerc I wasn't trying to be sarcastic or snarky, but anyway ...
The "welcoming" part in the title wasn't about the community but rather about the unwelcoming features of the language. When I first saw something about Carbon (from the designer of Odin), I was scared away by the syntax which looked just like the Rust syntax (I have struggled for about six months with an aborted Rust conversion--ironically it then ended being completed in Go which has similar Pascal-influenced declaration syntax). Later I watched your video with Titus Winters titled "What is C++?", and finally decided to take a second look by watching through your announcement video.
My point is that if you want to win the "hearts and souls" of the millions of C++ developers, I doubt that you'll be successful by using a syntax (not just for variable declarations) that is significantly different from the C++ syntax. Automated transition tools may convince a certain percentage of them, but not the bulk. That's my opinion, I may be wrong but I believe the history of programming languages tends to confirm my view.
My only other substantial point is *not* whether you want to claim the "successor" mantle or not, but rather what is written about Carbon principles on error handling. If Carbon is going to support object-oriented programming, I believe that Bjarne Stroustroup and Bertrand Meyer (see *Object-Oriented Software Construction*) would agree that some exception handling mechanism is required, rather than the Rust/Go approach. Returning error values may be appropriate for embedded environments but it's much easier to write and *read* code that is straightforward and is not sprinkled with error handling conditionals.

**Reply by geoffromer** [reply]
> If Carbon is going to support object-oriented programming, I believe that Bjarne Stroustroup and Bertrand Meyer (see _Object-Oriented Software Construction_) would agree that some exception handling mechanism is required, rather than the Rust/Go approach. Returning error values may be appropriate for embedded environments but it's much easier to write and _read_ code that is straightforward and is not sprinkled with error handling conditionals.

I think this is a false dichotomy. Implicitly-propagated exceptions and explicit error handling conditionals are not the only two ways of handling errors. See, for example, Rust's [`?` operator](https://doc.rust-lang.org/book/ch09-02-recoverable-errors-with-result.html#a-shortcut-for-propagating-errors-the--operator) and Swift's [`try` operator](https://docs.swift.org/swift-book/LanguageGuide/ErrorHandling.html), which are far more concise than explicit conditionals, but still explicitly mark the places in the code where error propagation happens.

**Reply by DevNoteHQ** [reply]
Well all new programming languages use this more script-like syntax. TypeScript, Kotlin, Switft, Rust,... I couldn't name a single modern language that sticks to the C++ syntax. So it seems to me that with the evolution of programming languages the C++ syntax has just been proven to be infavorable and thus the more script-like syntax gets used. Is it then really wise to stick to an outdated syntax just to be more welcoming to C++ devs? I mean that would be similar to using a deprecated feature just to make the code more readable for devs that are used to this exact feature. We are software developers. Dropping outdated features is what we get paid for^^

Also to share my personal experience: I mostly used C/C++ and a bit of Java in school. When I then got my first job, I had to switch to Kotlin and TypeScript. I thought that the syntax was weired in the beginning, but by using it in multiple projects so far I really favor it now over the C++/Java syntax. And I really hate the fact that I had to switch to Java for my current project.

**Reply by chandlerc** [reply]
> My point is that if you want to win the "hearts and souls" of the millions of C++ developers, I doubt that you'll be successful by using a syntax (not just for variable declarations) that is significantly different from the C++ syntax. Automated transition tools may convince a certain percentage of them, but not the bulk. That's my opinion, I may be wrong but I believe the history of programming languages tends to confirm my view.

As you say, time will tell.

We *are* trying to be mindful of the cost associated with the syntax changes. We may even re-evaluate those syntax decisions when we get to the point where we can precisely identify things that are creating enough problems for enough developers to be a critical issue to fix even if it makes things like parsing. But for now, we think we're making the best tradeoff we can given our experience and information.

**Reply by chandlerc** [reply]
> I couldn't name a single modern language that sticks to the C++ syntax

FWIW, Dart uses a pretty similar variable syntax, but of course has other differences.

**Reply by jmafc** [reply]
@geoffromer It may be more concise but it still places the onus on the programmer to do the error propagation.
@DevNoteHQ Just because a feature is not adopted by more "modern" languages doesn't mean it's better and the "older" one is "outdated" or "deprecated". But all this gets back to the basic question: Why not Rust? I'm pretty sure that Google is using it already (and Go and Dart). IIRC Go was introduced because Rob Pike couldn't stand the long compilation times of C++. If the objective is to port Google's large C++ codebase that is not already ported/being ported to Go or Rust, and the Carbon team is so enamored with the Rust style, why not put the effort into a transpilation effort from C++ to Rust (IIRC there is one already), instead of inventing yet another wheel, currently not well-defined but with *definite* tendencies towards the latter. What are the *benefits* to the general programming community of embarking in such an effort?

**Reply by jonmeow** [reply]
> But all this gets back to the basic question: Why not Rust?

Please see ["Why not Rust?" in the FAQ.](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/faq.md#why-not-rust)

**Reply by jmafc** [reply]
@jonmeow I saw it, but that's why I asked what are the benefits to the general programming community and specifically the C++ community. Other than Google and perhaps Microsoft, what other organizations have such large, complex C++ codebases and would *like* to move to something like Rust? I presume Google management, at some level, is agreeable to this project, but are there other corporate "sponsors"?

**Reply by DevNoteHQ** [reply]
> @DevNoteHQ Just because a feature is not adopted by more "modern" languages doesn't mean it's better and the "older" one is "outdated" or "deprecated".

It's not directly implied just by the amount of adoptions, yes. But modern languages have their reasons why they choose this syntax over the C++ syntax (easier parsability, better readability,...) and because so many come to the same conclusion, I would indeed call C++ syntax outdated.

**Reply by ahmet-voys** [reply]
> > I couldn't name a single modern language that sticks to the C++ syntax
> 
> FWIW, Dart uses a pretty similar variable syntax, but of course has other differences.

AFAIK, one of Dart Language designers' regret was to try too hard being `familiar`. I wish they could chime in with their experiences.

**Reply by OlaFosheimGrostad** [reply]
> AFAIK, one of Dart Language designers' regret was to try too hard being `familiar`. I wish they could chime in with their experiences.

You probably think of Gilad Bracha who expressed that the semantics of Dart were bland, but his dynamic language [Newspeak](https://en.wikipedia.org/wiki/Newspeak_(programming_language)) is a far cry from what Carbon is aiming for...

**Comment by jmafc** [comment]
@chandlerc I think it would be helpful for any C++ developer to see each C++ keyword (as in https://en.cppreference.com/w/cpp/keyword) and other key concepts listed in one place, together with the current Carbon determination as to what wiil be done about it, e.g., kept as is, kept but modified (Go example: `if` keyword is kept but parentheses are eliminated), replaced by (e.g., `int` apparently replaced by `i32` --but what is done if intent is a machine word), dropped, new syntax (e.g., the `fn` keyword), to be studied further, etc., including summary totals by category.

**Reply by DevNoteHQ** [reply]
I would agree on this but just in a later stage of development. Adding all differences to C++ now when basically the whole syntax could still change is just gonna be a nightmare to maintain.
For the summary do you mean something like this: [docs/design](https://github.com/carbon-language/carbon-lang/tree/trunk/docs/design)? Or a shorter version maybe?

**Reply by jmafc** [reply]
@DevNoteHQ The docs/design is certainly a start. [A relational database or spreadsheet would keep it from being "a nightmare to maintain".]

**Reply by geoffromer** [reply]
(I've edited the comment above to remove sarcasm, which our [code of conduct](https://github.com/carbon-language/carbon-lang/blob/trunk/CODE_OF_CONDUCT.md) doesn't permit. My edit is marked with square brackets.)

**Comment by jmafc** [comment]
First of all, @geoffromer please point out *where* does the Code of Conduct specify that sarcasm is not permitted. Although @chandlerc did indicate that "being sarcastic or snarky is not productive" and I did reply that I "wasn't trying to be sarcastic or snarky", he did not say there was a CoC prohibition, nor did he provide examples of what he found sarcastic in my original post.

Second, when I was replying to @DevNoteHQ's comment, I was about to write something like "You have heard of relational databases or spreadsheets, haven't you?", but I reflected on Chandler's earlier comment so instead I wrote "I was going to be snarky and ask whether you've heard of relational databases or spreadsheets to maintain such a list so that it isn't "a nightmare to maintain" but after advice from Chandler I won't :-)". In case it's not obvious, there is a *smiley* at the end of that, meaning I was trying to be friendly and *not* snarky, certainly not menacing or insulting in any way.

Curiously, Mathias gave my comment a thumbs-up, so he didn't seem offended (although perhaps he only saw it after your edit and removal of the smiley). Frankly, your blocking me, even if it was for only 24 hours, is at least the third instance of not being welcome by the Carbon community. Over the past decade I've interacted with many open source communities, e.g., PostgreSQL, Python, Eiffel, D, Go and Rust. I believe only one of them *explicitly* mentioned "welcoming environment" and coincidentally, that was the one for Rust.  However, Rust was the *one* community which I found most *unwelcoming*. For example, IIRC on a Discord discussion about how to do I/O in a DBMS server, someone essentially said that I was out of my mind if I wanted to use raw I/O rather than Rust's serialization methods. I'm afraid it's counterproductive to want or insist publicly that a group of people adopt a certain attitude. It's much easier to lead by example, as they say.

In closing, I think it would be beneficial for anyone to read and reflect on the following written by Bjarne Stroustrup in *The Design and Evolution of C++* (1994): "Many C++ design decisions have their roots in my dislike for forcing people to do things in some particular way. In history, some of the worst disasters have been caused by idealists trying to force people into 'doing what is good for them'" (p. 23). In fact, I think the entire section 1.3 (pp. 23-25) is enlightening.

Farewell.

**Reply by DevNoteHQ** [reply]
> Curiously, Mathias gave my comment a thumbs-up, so he didn't seem offended (although perhaps he only saw it after your edit and removal of the smiley).

That thumbs up was more meant towards "The docs/design is certainly a start.". And I've seen it before the edit^^
Though I have to say that sarcasm is my second language, so I didn't mind^^

> Many C++ design decisions have their roots in my dislike for forcing people to do things in some particular way. In history, some of the worst disasters have been caused by idealists trying to force people into 'doing what is good for them'

That's more applicable to politics then programming imo^^ I mean why would you even care for safety features in your language then? If the programmer wants to write horribly unsafe code, why shouldn't he be able to? Why should you use something like ESLint to enforce a certain "style" of code? In programming it's less idealists and more experienced developers that force certain things on you because they did it themselfes and know it will lead to bugs^^

**Reply by jmafc** [reply]
Indeed Mathias, Bjarne mentions "political themes" at the end of that section. But he also says things like
> However, a programming language is really a very tiny part of the world, and as such, it ought not be taken too seriously. Keep a sense of proportion and—most importantly—keep a sense of humor. Among major programming languages, C++ is the richest source of puns and jokes. That is no accident.

However, I doubt that Bjarne's remark implies that programmers should disregard safety or not care about horrible code. If you've read some of his other books or heard his talks, you know he cares deeply about those things.

**Comment by chandlerc** [comment]
Our code of conduct says to be welcoming, friendly, and patient, and in general, if someone asks you to stop, then stop.

Many members of the community do not find sarcasm or snark as part of arguing for a technical position welcoming or friendly. You were asked to avoid this, and instead included it. Whether it was intended or not, this is continuing to post things that others find unwelcoming and have asked you to stop.

Since it seems our community's approach isn't the one you want to follow, it probably isn't for you. FWIW, I greatly admire the Rust community and hope that Carbon will remain a similarly welcoming one.

Given this discussion isn't productive and has veered off the original topic, hiding the off-topic comments and leaving the thread locked.
